### PR TITLE
Add Windows ARM64 support

### PR DESCRIPTION
**Pull request by driver1998**
The code handling import thunks on ARM64 is based on the implementation of Detours (by Microsoft, MIT): https://github.com/microsoft/Detours/blob/734ac64899c44933151c1335f6ef54a590219221/src/detours.cpp#L1058.

All tests passed locally, on both MSVC and MinGW with Clang.

### CODE REVIEW

File: src/dlfcn.c [hunk 1]
```diff
@@ -680,6 +680,16 @@ static BOOL is_valid_address( const void *addr )
...
+static INT64 sign_extend(UINT64 value, UINT bits)
+{
+    const UINT left = 64 - bits;
+    const INT64 m1 = -1;
+    const INT64 wide = (INT64)(value << left);
+    const INT64 sign = (wide < 0) ? (m1 << left) : 0;
+    return value | sign;
+}
+
+
```
Comment (pali) [inline src/dlfcn.c:692]: Extra newline?

File: src/dlfcn.c [hunk 2]
```diff
@@ -689,7 +699,18 @@ static BOOL is_valid_address( const void *addr )
  */
 static BOOL is_import_thunk( const void *addr )
 {
+#if defined _M_ARM64 || defined __aarch64__
```
Comment (pali) [inline src/dlfcn.c:702]: Code style in this file is:
```suggestion
#if defined(_M_ARM64) || defined(__aarch64__)
```

File: src/dlfcn.c [hunk 3]
```diff
@@ -689,7 +699,18 @@ static BOOL is_valid_address( const void *addr )
  */
```
Comment (pali) [inline src/dlfcn.c:699]: I would suggest to extend comment because it describe x86 behavior.

File: src/dlfcn.c [hunk 4]
```diff
@@ -698,11 +719,28 @@ static BOOL is_import_thunk( const void *addr )
...
+     */
+    ULONG opCode1 = *(ULONG*)((BYTE*)addr);
+    ULONG opCode2 = *(ULONG*)((BYTE*)addr + 4);
+
+    UINT64 pageLow2 = (opCode1 >> 29) & 3;
+    UINT64 pageHigh19 = (opCode1 >> 5) & ~(~0ull << 19);
+    INT64 page = sign_extend((pageHigh19 << 2) | pageLow2, 21) << 12;
+
+    UINT64 offset = ((opCode2 >> 10) & ~(~0ull << 12)) << 3;
+    BYTE* ptr = (BYTE*)((ULONG64)thkp & 0xfffffffffffff000ull) + page + offset;
```
Comment (pali) [inline src/dlfcn.c:737]:
```suggestion
    BYTE *ptr = (BYTE *) ( (ULONG64) thkp & 0xfffffffffffff000ull ) + page + offset;
```
or maybe simplify `0xfffffffffffff000ull` by `~0xfffull`

**Review by pali — COMMENTED** [review]
Hello! Thank you very much for a change. It looks good, just a few small suggestions to following existing coding style in this file for consistency.

**Comment by pali** [comment]
Hello! I see that on CI appeared a new warning from this pull request:
```
[  8%] Building C object src/CMakeFiles/dl.dir/dlfcn.c.obj
C:\projects\dlfcn-win32\src\dlfcn.c:683:14: warning: 'sign_extend' defined but not used [-Wunused-function]
 static INT64 sign_extend(UINT64 value, UINT bits)
              ^
```
So please fix it. And also please look at other changes related to code style (spaces in is_import_thunk and get_address_from_import_address_table) to have consistency.

**Comment by driver1998** [comment]
Sorry for the late reply, patches are updated again, please review.

**Comment by pali** [comment]
Looks good for me! Squash fixup commits into one and then PR would be ready for merge.

@traversaro Any other comments?

**Comment by driver1998** [comment]
Should be done now, (messed up with line endings for a few commits).

**Comment by traversaro** [comment]
> Looks good for me! Squash fixup commits into one and then PR would be ready for merge.
> 
> @traversaro Any other comments?

No! Thanks a lot @driver1998 and @pali for the work. Do you think there will be any follow up work or I can proceed with a release after this PR is merged?

**Pull request merged**